common/src/main/java/io/netty/util/internal/ThreadCleaner.java

```diff
@@ -0,0 +1,131 @@
+                    try {
+                        ThreadCleanerReference reference =
+                                (ThreadCleanerReference) REFERENCE_QUEUE.remove();
+                        try {
+                            reference.cleanup();
+                        } finally {
+                            LIVE_SET.remove(reference);
+                        }
+                    } catch (InterruptedException ex) {
+                        // Just consume and move on
+                        interrupted = true;
+                    }
+                }
+                CLEANER_RUNNING.set(false);
+
+                // Its important to first access the LIVE_SET and then CLEANER_RUNNING to ensure correct
+                // behavior in multi-threaded environments.
+                if (LIVE_SET.isEmpty() || !CLEANER_RUNNING.compareAndSet(false, true)) {
+                    // There was nothing added after we set STARTED to false or some other cleanup Thread
+                    // was started already so its safe to let this Thread complete now.
+                    break;
+                }
+            }
+            if (interrupted) {
+                // As we catched the InterruptedException above we should mark the Thread as interrupted.
+                Thread.currentThread().interrupt();
+            }
+        }
+    };
+
+    /**
+     * Register the given {@link Thread} for which the {@link Runnable} will be executed once there are no references
+     * to the object anymore, which typically happens once the {@link Thread} dies.
+     *
+     * This should only be used if there are no other ways to execute some cleanup once the {@link Thread} dies as
+     * it is not a cheap way to handle the cleanup.
+     */
+    public static void register(Thread thread, Runnable cleanupTask) {
+        ThreadCleanerReference reference = new ThreadCleanerReference(thread,
+                ObjectUtil.checkNotNull(cleanupTask, "cleanupTask"));
+        // Its important to add the reference to the LIVE_SET before we access CLEANER_RUNNING to ensure correct
+        // behavior in multi-threaded environments.
+        LIVE_SET.add(reference);
+
+        // Check if there is already
+        if (CLEANER_RUNNING.compareAndSet(false, true)) {
+            Thread cleanupThread = new Thread(CLEANER_TASK);
+            cleanupThread.setPriority(Thread.MIN_PRIORITY);
+            // Set to null to ensure we not create classloader leaks by holds a strong reference to the inherited
+            // classloader.
+            // See:
+            // - https://github.com/netty/netty/issues/7290
+            // - https://bugs.openjdk.java.net/browse/JDK-7008595
+            cleanupThread.setContextClassLoader(null);
+            cleanupThread.setName("ThreadCleanerReaper");
+
+            // This Thread is not a daemon as it will die once all references to the registered Threads will go away.
+            cleanupThread.setDaemon(false);
```

Comment:
Daemon is for not blocking shutdown.  In a shutdown scenario, how important is it to clean up refs?

Reply:
actually I think we could also just use a non-daemon thread as the thread will go away once there is no references anymore.